File: src/trustrag/extract/pdf.py
"""PdfExtractor — text per page with page numbers + word bboxes via pdfplumber (§8)."""

from __future__ import annotations

from typing import Any

import pdfplumber

from trustrag.evidence.unit import BBox
from trustrag.extract.plain import open_binary
from trustrag.extract.types import (
    BlockKind,
    ExtractedBlock,
    ExtractedDoc,
    ImageRef,
    SourceType,
    StructureType,
)
from trustrag.plugins.base import ExtractorPlugin


def _page_text_bbox(page: Any) -> BBox | None:
    """The bounding box covering every extracted word on the page, where feasible."""
    words = page.extract_words()
    if not words:
        return None
    x0 = min(float(w["x0"]) for w in words)
    top = min(float(w["top"]) for w in words)
    x1 = max(float(w["x1"]) for w in words)
    bottom = max(float(w["bottom"]) for w in words)
    return (x0, top, x1, bottom)


class PdfExtractor(ExtractorPlugin):
    """One paragraph block per page (text + page number + a word-derived bbox);
    page images become ``ImageRef``s anchored to their page and box."""

    plugin_version = "pdf/1"

    def __init__(self, document_id: str | None = None) -> None:
        self.document_id = document_id

    def extract(self, source: Any) -> ExtractedDoc:
        opened, doc_id, source_uri = open_binary(source, self.document_id)

        blocks: list[ExtractedBlock] = []
        images: list[ImageRef] = []
        with pdfplumber.open(opened) as pdf:
            for index, page in enumerate(pdf.pages):
                number = index + 1
                text = (page.extract_text() or "").strip()
                blocks.append(
                    ExtractedBlock(
                        order=index,
                        kind=BlockKind.paragraph,
                        text=text,
                        page=number,
                        bbox=_page_text_bbox(page),
                    )
                )
                for img_index, img in enumerate(page.images):
                    images.append(
                        ImageRef(
                            image_id=f"page{number}-img{img_index}",
                            page=number,
                            bbox=(
                                float(img["x0"]),
                                float(img["top"]),
                                float(img["x1"]),
                                float(img["bottom"]),
                            ),
                        )
                    )

        return ExtractedDoc(
            document_id=doc_id,
            source_type=SourceType.pdf,
            structure_type=StructureType.page_layout,
            source_uri=source_uri,
            blocks=tuple(blocks),
            images=tuple(images),
        )

File: rust/src/extract/pdf.rs
//! PdfExtractor — text per page + page-level bbox, via pdfium (feature `pdf`).
//!
//! Mirrors `extract/pdf.py`: one paragraph block per page (1-based `page`),
//! text from the page's text layer, and each page image as an
//! `ImageRef(page{N}-img{i})`. pdfium-render binds libpdfium dynamically at
//! runtime, so the crate compiles without it; enable the feature and ship
//! the pdfium library alongside (see pdfium-render docs).

use pdfium_render::prelude::*;

use crate::types::*;

pub fn extract(
    bytes: &[u8],
    document_id: &str,
    source_uri: Option<String>,
) -> Result<ExtractedDoc, String> {
    let pdfium = Pdfium::default();
    let document = pdfium
        .load_pdf_from_byte_slice(bytes, None)
        .map_err(|e| format!("pdfium: {e}"))?;

    let mut blocks = Vec::new();
    let mut images = Vec::new();

    for (index, page) in document.pages().iter().enumerate() {
        let number = (index + 1) as u32;
        let text = page
            .text()
            .map(|t| t.all())
            .unwrap_or_default()
            .trim()
            .to_string();
        blocks.push(ExtractedBlock {
            order: index,
            kind: BlockKind::Paragraph,
            text,
            page: Some(number),
            bbox: None, // word-derived bbox lands with the conformance pass
            level: None,
            structure_path: vec![],
        });

        let mut img_index = 0usize;
        for object in page.objects().iter() {
            if object.as_image_object().is_some() {
                let bounds = object.bounds().ok();
                images.push(ImageRef {
                    image_id: format!("page{number}-img{img_index}"),
                    page: Some(number),
                    bbox: bounds.map(|b| {
                        [
                            b.left().value as f64,
                            b.top().value as f64,
                            b.right().value as f64,
                            b.bottom().value as f64,
                        ]
                    }),
                    width: None,
                    height: None,
                    blob_key: None,
                });
                img_index += 1;
            }
        }
    }

    Ok(ExtractedDoc {
        document_id: document_id.to_string(),
        source_type: SourceType::Pdf,
        structure_type: StructureType::PageLayout,
        source_uri,
        blocks,
        images,
    })
}
